### PR TITLE
Iterator for diskann

### PR DESCRIPTION
**Pull request by alwayslove2013**
- add iterator support for knowhere diskann
- remove diskann-range_search, use index_node-range_search instead. (implemented by diskann-iterator)

issue: https://github.com/milvus-io/milvus/issues/34816
co-author: @xxxlzhxxx

### CODE REVIEW

File: thirdparty/DiskANN/src/pq_flash_index.cpp [hunk 1]
```diff
@@ -58,6 +58,130 @@ namespace {
...
+    }
+
+    if (metric == diskann::Metric::INNER_PRODUCT ||
+        metric == diskann::Metric::COSINE) {
+      query_norm = std::sqrt(query_norm);
+      if (metric == diskann::Metric::INNER_PRODUCT) {
+        aligned_query_T[q_dim] = 0;
+        aligned_query_float[q_dim] = 0;
+      }
+      for (uint32_t i = 0; i < q_dim; i++) {
```
Comment (alexanderguzhva) [inline thirdparty/DiskANN/src/pq_flash_index.cpp:101]: missing `if (metric == diskann::Metric::COSINE)` ?

File: thirdparty/DiskANN/src/pq_flash_index.cpp [hunk 2]
```diff
@@ -58,6 +58,130 @@ namespace {
...
+        aligned_query_T[i] = (T) ((float) aligned_query_T[i] / query_norm);
```
[the remark this answers is not in the document]
Comment (alwayslove2013) [reply thirdparty/DiskANN/src/pq_flash_index.cpp]: fixed.

File: thirdparty/DiskANN/src/pq_flash_index.cpp (hunk 1 above)
Comment (alwayslove2013) [reply thirdparty/DiskANN/src/pq_flash_index.cpp]: There is a special calculation method for `IP` metric, and an extra dim (the last) will be used for “convert MIPS to L2 search”.

File: thirdparty/DiskANN/src/pq_flash_index.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexanderguzhva) [reply thirdparty/DiskANN/src/pq_flash_index.cpp]: @alwayslove2013 
```C++
    aligned_query_T[q_dim]
    aligned_query_float[q_dim]
```
For the fixed version, this two elements do not get populated for `query_norm == 0` use case.

[the remark this answers is not in the document]
Comment (alexanderguzhva) [reply thirdparty/DiskANN/src/pq_flash_index.cpp]: Basically,
```C++
    if (query_norm != 0) {
        query_norm = std::sqrt(query_norm);
        if (metric == diskann::Metric::INNER_PRODUCT) {
          aligned_query_T[q_dim] = 0;
          aligned_query_float[q_dim] = 0;
        }
        for (uint32_t i = 0; i < q_dim; i++) {
          aligned_query_T[i] = (T) ((float) aligned_query_T[i] / query_norm);
          aligned_query_float[i] /= query_norm;
        }
      }
    else {
      .. do what.. ?
    }
```

[the remark this answers is not in the document]
Comment (alwayslove2013) [reply thirdparty/DiskANN/src/pq_flash_index.cpp]: No need to do anything.
When the metric is `ip` or `cosine` and `query_norm==0`, `workspace next batch` will be skipped. iterator's results will not be updated and will always be empty, which means `Iterator.HasNext()` will be `false`.
```
void PQFlashIndex<T>::getIteratorNextBatch(IteratorWorkspace<T> *workspace) {
  if (metric == diskann::Metric::INNER_PRODUCT ||
      metric == diskann::Metric::COSINE) {
    if (workspace->query_norm == 0) {
      return;
    }
  }
  ...
}
```
I will modify the logic to make it clearer.

File: src/index/diskann/diskann.cc [hunk 3]
```diff
@@ -152,7 +149,41 @@ class DiskANNIndexNode : public IndexNode {
         return knowhere::IndexEnum::INDEX_DISKANN;
     }
 
+    expected<std::vector<IndexNode::IteratorPtr>>
+    AnnIterator(const DataSetPtr dataset, std::unique_ptr<Config> cfg, const BitsetView& bitset) const override;
+
  private:
+    class iterator : public IndexIterator {
```
Comment (PwzXxm) [inline src/index/diskann/diskann.cc:156]: Personally, I will naming this staring with an uppercase

File: src/index/diskann/diskann.cc [hunk 4]
```diff
@@ -520,6 +551,58 @@ DiskANNIndexNode<DataType>::Deserialize(const BinarySet& binset, std::shared_ptr
...
+                                                                                "unsupported metric type");
+    }
+
+    constexpr uint64_t k_lsearch_iterator = 32;
+    auto lsearch = static_cast<uint64_t>(search_conf.search_list_size.value_or(k_lsearch_iterator));
+    auto beamwidth = static_cast<uint64_t>(search_conf.beamwidth.value());
+    auto filter_ratio = static_cast<float>(search_conf.filter_threshold.value());
+
+    bool use_reorder_data = false;
+    bool for_tuning = false;
```
Comment (PwzXxm) [inline src/index/diskann/diskann.cc:576]: these two default to `false`? Possibly pass them in from config?

Comment (alwayslove2013) [reply src/index/diskann/diskann.cc]: both removed.

**Pull request merged**